Add further functionality to iex-rs

jakeschurch opened the pull request:
Refactor majority of code to allow for flexibility, maintainability, and ease-of-use.

## Features
- Add Request enum to replace all client methods to one request method 
- Use serde_json::Value to avoid different method signature return values for request method
- Refactor Client struct signatures
- Add derive trait statements to structs & enums as needed
- Implement chart & duration-required Request enums
- Add code to ToString trait for Request
- Add helper emus ListParam & ChartParam for Request enum
- Add parse_params fn for creating url chart query
- Refactor test suite
- Update & add labelled comments - please answer in comments section if it seems like response needed!

Note to @samwho :
Interested in helping maintain/contribute to this repo. Let me know if I can ever help with anything.

Commit:
Refactor code for more modularity & readability

- Implement Request enum to replace N Request funcs -> 1
- Refactor Client struct signatures
- Implement range arg for Request::News
- Start Refactor of test suite

Commit:
Finish first stable draft of new features

- Add derive trait statements to some structs
- Implement chart & duration-required Request enums
- Add code to ToString trait for Request
- Add helper structs ListParam & ChartParam for Request enum
- Add parse_params fn for creating url chart query
- Finish refactor of test suite
- Update & add labelled comments - please answer if needs feedback

Commit:
Merge branch 'feature/iex-request-enum' into develop

Refactor majority of code to allow for future-proofing, readability, and ease-of-use.

samwho commented:
Thanks for this PR! I’ll try and review it over the weekend. :)

samwho commented on src/lib.rs:
```diff
@@ -54,347 +58,456 @@ impl Default for Duration {
-    pub chart_interval: Option<i64>,
+#[derive(PartialEq, Eq)]
+/// The `Request` enum type allows for HTTP request matching to the IEX API.
+// TODO(Request): Add documentation from IEX website.
```
I think just a link to the relevant part of the documentation would suffice. No need to add it in this PR, though.

samwho commented on src/lib.rs:
```diff
@@ -54,347 +58,456 @@ impl Default for Duration {
+    Book {
+        symbol: &'a str,
+    },
+    // REVIEW(CHART): Should `time-series` be implemented even though it forwards
```
Yeah, fo sho. I wanted to mirror what was available on IEX 1 to 1.

samwho commented on src/lib.rs:
```diff
@@ -54,347 +58,456 @@ impl Default for Duration {
-impl Default for ChartParams {
-    fn default() -> ChartParams {
-        ChartParamsBuilder::default().build().unwrap()
+impl<'a> ToString for Request<'a> {
```
I'm not 100% on this implementation, because it makes you jump around to figure out how URIs are formed.

I preferred the approach of having a method per API endpoint and putting all of the logic in each of those methods. What benefit does this proposed new approach have over the old one?

jakeschurch replied:
@samwho I'm not sure that I am on the same page as you are. If you are referring specifically to the building of the implementation of the chart params url then I can easily take away the parse params function and insert the logic into the ChartParms block of the Request ToString method. Regarding specifically why I did this, is because at the time it made sense to follow the logic of the Duration enum. Also, the method does not need to be exposed to the public. If I did not address your question, please let me know. 

---

Regarding the benefits of just one API endpoint: 
- We are exposing less to the end-user - (benefit on its own)
-  The multiple endpoints method signatures had the same behaviour, which was to request a Response - whether that would be dividends, earnings, etc. I do not think that it is idiomatic to expose intentions of state for a method, instead of behaviour. With the multiple-endpoint API, we expose state, and hide behaviour. Meaning that dividends (inferring state) is a `pub fn`, but `get` which infers behaviour, is hidden. 
    - However, with a single endpoint the method signatures are the same. The multiple endpoints repeat the same behaviour with different inputs (state variables), and all we are doing with the request enum is coalescing all of the multiple endpoint's states into one type, which all have the same behaviour. 
- This helps maintainability because intentions are inherently as well as easily expressed.
- Method signatures are the same. We return a serde_json::Value instead of a custom type. Since there is no behaviour wrapped around the structs found in types.rs, there is no reason to return the value as the specific type when it has the same functionality and behavior as serde_json::Value. We leave it up the end-user if they want to change the return struct's type. 
-  Having the type-variant-enum approach provides bounds for maintainability. Instead of creating another method, we just add to our request enum, and update the ToString method
    -  Have you thought about how you are going to implement batch requests? Allowing a variadic Request argument to this particular method is going to make implementation extremely easy. Personally, I am not sure how batch requests would be implemented with multiple endpoints. First off, what would be the input arguments? Nested methods or a variadic method signature is not exactly elegant. Personally, I would suggest a variadic argument consisting of Request enum type variants. This way, we can worry about state, and expose behaviour. However, if we stick with multiple endpoints, you can either use serde_json::Value as a return type wrapped in a result, or, we would need to create n! ~2^n~ methods ~signatures~ for each mix and match of the single state methods - which I think you would agree, should not be used for batch requests.

I think this is enough to express the majority of my ideas for the single-point implementation, but if need be, I can provide more input.

samwho replied:
That makes sense.

I like to try and make the surface of APIs that are mostly wrapping calls to HTTP endpoints look like those endpoints as much as possible, but agree that this approach has merit and am happy to go forward with it.

samwho commented on src/lib.rs:
```diff
@@ -54,347 +58,456 @@ impl Default for Duration {
+    Interval(usize),
+    /// boolean. If true, changeOverTime and marketChangeOverTime will be relative to previous day close instead of the first value.
+    ChangeFromClose(bool),
+    /// number. If passed, chart data will return the last N elements.
```
Remove the type in each of the comments. They're redundant.

samwho commented on src/lib.rs:
```diff
@@ -54,347 +58,456 @@ impl Default for Duration {
-        // scenario.
-        self.get(&format!("/stock/{}/previous", symbol))
-    }
+    // NOTE(test): Unfortunately rust has no implementation of sub-testing, or at least not aware of as of yet.
```
Not quite sure what this comment means...

jakeschurch replied:
Just informational. Originally I thought it would be best to take all of the tests that I have implemented in the PR and have them under one function. However, when I first ran that specific function, it did not tell me which assert failed, which is why I broke that one function into 20+

jakeschurch commented:
Hey @samwho,

Since you were not 100% on my implementation of the new singular request system, I decided to try a different approach. 

I started drafting the implementation, and just due to the overall time commitment of implementing this new version, I would like to get your approval first before fully implementing/drafting it. You can see the overall [gist here.](https://gist.github.com/jakeschurch/1af302b68c5ad96bd1526d052cc82708#file-client_batch-rs)

Major Changes:

- client uses builder pattern to create a request:
```rust
Client::new().types(vec![Request::Foo]).symbols(vec!["Bar", "buz"]).request();
```
- The benefit to this new approach is the way in which we construct URLs. Ultimately, every url that is constructed will follow the batch request pattern allowing us to access singular requests as well as multiple in the batch request format.

    - Examples:
        - Single request: `https://api.iextrading.com/1.0/stock/market/batch?symbols=aapl&types=quote`
        - _Actual_ batch request: `https://api.iextrading.com/1.0/stock/market/batch?symbols=aapl,fb&types=quote,news`

    - This approach will allow us to also add parameters on the end, which is automatically pointed to the correct endpoint which is helpful.

- I haven't exactly decided how parameters should be taken care of. Obviously the Request enum needs to be updated to reflect all of these changes. First off I know that the enum will no longer have a symbol field. That leaves the rest of the fields, mainly reflecting duration, ChartParams, and ListParams. I can extract these from the specific types, or, I can introduce a params method & Vec to collect and hold them -- your call.

Let me know how you feel about the main idea or if you have any questions. Thanks!

**Side Note:** I have started working on iextp & TOPS protocol so that we can make historical requests/parse pcap data. Hopefully I can get the majority flushed out within the next week or so.

Best,
@jakeschurch

samwho commented:
Just dropping a note to say I haven’t forgotten about this, and will give it the time it deserves when I can.

Thanks again for your contributions, I do really appreciate them. :)

samwho commented:
Not an enormous fan of using the batch endpoint for everything. Thinking over your initial approach in this PR I think it's sound and happy for you to press on with it. :)

I've invited you to be a collaborator, instead of having to wait for me to chime in.

Commit:
Implement base functionality for IEX Stocks Endpoints.

- Remove parse_params fn
- Change name of Request enum to StocksEndpoint
- Move anything largely dealing with endpoints to endpoints.rs
- Update test suite
- Remove type in each of comments
- Add Endpoint Trait
- add other minimal changes

jakeschurch commented:
I sincerely appreciate that @samwho!
If there is anything that you would like my involvement in implementing, please let me know through an issue or a PR. 
Do you have any specific guidelines for me acting as a collaborator? If not, I'll just make sure to follow fairly-straightforward gitflow guidelines. I know the general goals for this repo, and plan on acting on those.
